### ISSUE TITLE
Embeddable with all fields set to `null` is hydrated as `undefined`

### ISSUE THREAD

**Issue by DASPRiD** [mikro-orm/mikro-orm#2774]
**Describe the bug**
When an embeddable has all its fields set to `null`, it will be hydrated as `undefined` upon retrieval.

**To Reproduce**
@Embeddable()
class Bar {
  @Property({type: t.string, nullable: true})
  public value : string | null;
}

@Entity()
class Foo {
    @PrimaryKey({type: t.uuid})
    public readonly id : string;

    @Embedded(() => Bar)
    public bar : Bar;
}

1. Create and persist `Foo` entity with `Bar` embeddable value set to `null`.
2. Clear the identity map.
3. Retrieve entity back from the entity manager.
4. `foo.bar.value === null` fails, because `foo.bar` is `undefined`.

**Expected behavior**
`foo.bar.value === null` should succeed.

**Versions**

| Dependency | Version |
| - | - |
| node | 1.16 |
| typescript | 4.4.3 |
| mikro-orm | 5.0.1 |
| your-driver | postgres |

**Comment by B4nan** [comment]
Not reproducible. Maybe you are using object mode embeddables? In such case, you need to explicitly set the `null` value, otherwise it gets lots when the value is serialized to JSON.

**Comment by DASPRiD** [comment]
Can you point me to the file where the hydration for embeddables takes place? I have a really hard time finding my way around the codebase. If I knew where to look, I could debug it myself :)

**Comment by B4nan** [comment]
Are you using object embeddables or not?

**Comment by DASPRiD** [comment]
I'm not.

**Comment by B4nan** [comment]
Then please provide a failing repro. That will always be the first step, I need to see things failing, won't be accepting PRs without tests anyway :]

Hydration happens in ObjectHydrator, its done via compiled functions.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts#L208

**Comment by DASPRiD** [comment]
Thanks, I'll have a look at this over the course of this week and will provide at least a failing test case :)

**Comment by B4nan** [comment]
Note that the hydration works via the exact same method as for regular columns - in the end they are regular columns, just mapped to a nested object structure.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts#L267

Only case where we explicitly set things to `undefined` is when you use `forceUndefined` configuration:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts#L290-L292

The value should be already `null` from the database, this is not handled via the ORM anyhow.

**Comment by DASPRiD** [comment]
I PRd a failing test case for this issue.

From what I can see here, this might actually be intended behavior:
https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts#L217-L223

If that's the case, there should potentially be an option for the `Embedded` decorator to disable that.

**Comment by DASPRiD** [comment]
Right, although in my instance, the `name` property itself should never be null (my constructor and its typings don't allow that). So yeah, would be great to have that configurable :)

**Comment by B4nan** [comment]
I guess we can make this work based on the nullability of the embedded property, makes sense.

**Comment by DASPRiD** [comment]
That makes a lot of sense to me, yeah, it's not used otherwise really :+1:

**Comment by B4nan** [comment]
All properties are non-nullable by default, so `@Property({nullable: false})` is the same as `@Property()`, it needs to work the same way for embeddables too, it just wan't taken into account.
